components/core/src/reducer/reducer_server.cpp

```diff
@@ -0,0 +1,504 @@
+
+    std::shared_ptr<ServerContext> ctx;
+    tcp::socket socket;
+    char* buf;
+    size_t buf_size;
+    size_t bytes_occupied{0};
+};
+
+int RecordReceiverContext::try_read_connection_init_packet(size_t num_bytes_read) {
+    bytes_occupied += num_bytes_read;
+
+    int32_t job_id{0};
+
+    if (bytes_occupied > sizeof(job_id)) {
+        SPDLOG_ERROR("Rejecting connection because of invalid negotiation");
+        return -1;
+    }
+    if (bytes_occupied < sizeof(job_id)) {
+        return 0;
+    }
+
+    memcpy(&job_id, buf, sizeof(job_id));
+    if (job_id != ctx->get_job_id()) {
+        SPDLOG_ERROR(
+                "Rejecting connection from worker with job_id={} during processing of "
+                "job_id={}",
+                job_id,
+                ctx->get_job_id()
+        );
+        return -1;
+    }
+    bytes_occupied = 0;
+
+    return 1;
+}
+
+bool RecordReceiverContext::try_read_record_groups_packet(size_t num_bytes_read) {
+    bytes_occupied += num_bytes_read;
+
+    size_t record_size{0};
+    auto* read_head = buf;
+    while (bytes_occupied > 0) {
+        if (bytes_occupied < sizeof(record_size)) {
+            break;
+        }
+        memcpy(&record_size, read_head, sizeof(record_size));
+
+        // terminate if record group size is over 16MB
+        if (record_size >= cMaxRecordSize) {
+            SPDLOG_ERROR("Record too large: {}B", record_size);
+            return false;
+        }
+
+        if (bytes_occupied < record_size + sizeof(record_size)) {
+            break;
+        }
+        read_head += sizeof(record_size);
+
+        auto record_group = DeserializedRecordGroup{read_head, record_size};
+        ctx->push_record_group(record_group.get_tags(), record_group.record_iter());
+        bytes_occupied -= (record_size + sizeof(record_size));
+        read_head += record_size;
+    }
+
+    if (bytes_occupied > 0) {
+        if (buf_size < record_size + sizeof(record_size)) {
+            auto new_buf = new char[record_size + sizeof(record_size)];
+            memcpy(new_buf, read_head, bytes_occupied);
+            delete[] buf;
+            buf = new_buf;
+        } else {
+            memmove(buf, read_head, bytes_occupied);
+        }
+    }
+
+    return true;
+}
+
+void queue_accept_task(std::shared_ptr<ServerContext> const& ctx);
+void queue_receive_task(std::shared_ptr<RecordReceiverContext> const& rctx);
+void queue_validate_sender_task(std::shared_ptr<RecordReceiverContext> const& rctx);
+void queue_scheduler_update_listener_task(
+        std::shared_ptr<ServerContext>& ctx,
+        size_t current_buffer_occupancy
+);
+
+struct ReceiveTask {
+    explicit ReceiveTask(std::shared_ptr<RecordReceiverContext> rctx) : rctx(std::move(rctx)) {}
+
+    void operator()(boost::system::error_code const& error, size_t num_bytes_read) {
+        // if no new bytes terminate
+        if (0 == num_bytes_read || ServerStatus::Running != rctx->ctx->get_status()) {
```

Comment:
Don't we need to check for `error.failed()` here like we do in ValidateSenderTask?

Reply:
Actually the reverse -- should also take it out there or change how we handle the error. The stream being closed will make `error.failed()` true even if there are bytes left over, so I was finding bytes==0 to be a better indicator.

Reply:
Actually nevermind, because of how the back and forth for connection validation works it is fine to exit on stream closed during validation.